vdev_disk: rewrite BIO filling machinery to avoid split pages

robn opened the pull request:
### Motivation and Context

This PR tackles a number of issues in the way BIOs (`struct bio`) are constructed for submission to the Linux block layer.

#### BIO segment limits are set incorrectly

The kernel has a hard upper limit on the number of pages/segments that can be added to a BIO, as well as a separate limit for each device (related to its queue depth and other scheduling characteristics).

ZFS counts the number of memory pages in the request ABD (`abd_nr_pages_off()`, and then uses that as the number of segments to put into the BIO, up to the hard upper limit. If it requires more than the limit, it will create multiple BIOs.

Leaving aside the fact that page count method is wrong (see below), not limiting to the device segment max means that the device driver will need to split the BIO in half. This is alone is not necessarily a problem, but it interacts with another issue to cause a much larger problem.

#### BIOs are filled inefficiently

The kernel function to add a segment to a BIO (`bio_add_page()`) takes a `struct page` pointer, and offset+len within it. `struct page` represents a run of contiguous memory pages (known as a "compound page"). In can be of arbitrary length.

The ZFS functions that count ABD pages and load them into the BIO (`abd_nr_pages_off()`, `bio_map()` and `abd_bio_map_off()`) will never consider a page to be more than `PAGE_SIZE` (4K), even if the `struct page` is for multiple pages. In this case, it will load the same `struct page` into the BIO multiple times, with the offset adjusted each time.

With a sufficiently large ABD, this can easily lead to the BIO being entirely filled much earlier than it could have been. This is also further contributes to the problem caused by the incorrect segment limit calculation, as its much easier to go past the device limit, and so require a split.

Again, this is not a problem on its own.

#### Incomplete pages are submitted to BIOs

The logic for "never submit more than `PAGE_SIZE`" is actually a little more subtle. It will actually never submit a buffer that crosses a 4K page boundary.

In practice, this is fine, as most ABDs are scattered, that is a list of complete 4K pages, and so are loaded in as such.

Linear ABDs are typically allocated from slabs, and for small sizes they are frequently not aligned to page boundaries. For example, a 12K allocation can span four pages, eg:

     -- 4K -- -- 4K -- -- 4K -- -- 4K --
    |        |        |        |        |
          :## ######## ######## ######:    [1K, 4K, 4K, 3K]

Such an allocation would be loaded into a BIO as you see:

    [1K, 4K, 4K, 3K]

This tends not to be a problem in practice, because even if the BIO were filled and needed to be split, each half would still have either a start or end aligned to the logical block size of the device (assuming 4K at least).

---

In ideal circumstances, these shortcomings don't cause any particular problems. Its when they start to interact with other ZFS features that things get interesting.

#### Aggregation

Aggregation will create a "gang" ABD, which is simply a list of other ABDs. Iterating over a gang ABD is just iterating over each ABD within it in turn.

Because the segments are simply loaded in order, we can end up with uneven segments either side of the "gap" between the two ABDs. For example, two 12K ABDs might be aggregated and then loaded as:

    [1K, 4K, 4K, 3K, 2K, 4K, 4K, 2K]

Should a split occur, each individual BIO can end up either having an start or end offset that is not aligned to the logical block size, which some drivers (eg SCSI) will reject. However, this tends not to happen because the default aggregation limit usually keeps the BIO small enough to not require more than one split, and most pages are actually full 4K pages, so hitting an uneven gap is very rare anyway.

#### Gang blocks

If the pool is under particular memory pressure, then an IO can be broken down into a "gang block", a 512-byte block composed of a header and up to three block pointers. Each points to a fragment of the original write, or in turn, another gang block, breaking the original data up over and over until space can be found in the pool for each of them.

Each gang header is a separate 512-byte memory allocation from a slab, that needs to be written down to disk. When the gang header is added to the BIO, its a single 512-byte segment.

#### Aggregation with gang blocks

Pulling all this together, consider a large aggregated write of gang blocks. This results a BIO containing lots of 512-byte segments. Given our tendency to overfill the BIO, a split is likely, and most possible split points will yield a pair of BIOs that are misaligned. Drivers that care, like the SCSI driver, will reject them.

---

### Description

This commit is a substantial refactor and rewrite of much of `vdev_disk` to sort all this out.

#### Configure maximum segment size for device

`vdev_bio_max_segs()` now returns the ideal maximum size for the device, if available. There's also a tuneable `vdev_disk_max_segs` to override this, to assist with testing.

#### ABDs checked up front for page count and alignment

We scan the ABD up front to count the number of pages within it, and to confirm that if we submitted all those pages to one or more BIOs, it could be split at any point with creating a misaligned BIO.

If that wouldn't be possible (as in any of the above situations), the ABD is linearised, and then checked again. This is the same technique used in `vdev_geom` on FreeBSD, adjusted for Linux's variable page size and allocator quirks.

In the end, a count of segments is produced, and this is combined with the max seg count to determine how many BIOs will be needed.

#### Virtual block IO object

`vbio_t` is a cleanup and enhancement of the old `dio_request_t`. The idea is simply that it can hold all the state needed to create, submit and return multiple BIOs, including all the refcounts, the ABD copy if it was needed, and so on. Apart from what I hope is a clearer interface, the major difference is that because we know how many BIOs we'll need up front, we don't need the old overflow logic that would grow the BIO array, throw away all the old work and restart. We can get it right from the start.

#### Other cleanups

Lots of cleanup, particularly through `vdev_disk_io_start` to make it feel a bit more like current ZFS code. Loads of comments too.

Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.

### How Has This Been Tested?

Full test suite run passed.

Various customer workloads have been tried (in QA/test environments), including a heavy write workload on 7x raidz3-12 pools. and are not showing any signs of problems, either from within ZFS or from block devices or elsewhere.

(Surprisingly, heavy async write loads are getting ~2.7x throughput, though sync write loads are a more modest ~1.1x. I had expected some improvement, as we're not spending so much time breaking down the ABDs, but I didn't imagine the overheads could be that high).

On stock ZFS the splitting issue is complicated to reproduce, as we need to:

- generate lots of large IOs;
- composed of small allocations;
- so as to raise the possibility that (a) a bio will be split by the kernel (b) at a point that will cause misalignment
- with the device driver or other stage below the block layer actually care about these misalignments and report on them.

The last one means the SCSI driver, unless there’s another I don’t know about.

The first three can be demonstrated by forcing all allocations to be gang blocks.

```
echo 32768 > /sys/module/zfs/parameters/metaslab_force_ganging
echo 100 > /sys/module/zfs/parameters/metaslab_force_ganging_pct

zpool create -o ashift=12 -O recordsize=1M -O compression=off tank raidz1 sda sdb ...

fio --name=gang --directory=/tank --ioengine=pvsync2 --rw=randrw --nrfiles=32 --filesize=1-8M --bsrange=1k-4k --runtime=60 --time_based
```

Without this PR, this test should yield `request not aligned to the logical block size` errors almost immediately:

```
Jul 19 00:26:26 ER6-U35 kernel: sd 0:0:10:0: [sdj] tag#3807 request not aligned to the logical block size
Jul 19 00:26:26 ER6-U35 kernel: blk_update_request: I/O error, dev sdj, sector 973441336 op 0x1:(WRITE) flags 0x4700 phys_seg 128 prio class 0
Jul 19 00:26:26 ER6-U35 kernel: sd 0:0:10:0: [sdj] tag#3833 request not aligned to the logical block size
Jul 19 00:26:26 ER6-U35 kernel: blk_update_request: I/O error, dev sdj, sector 973442332 op 0x1:(WRITE) flags 0x700 phys_seg 13 prio class 0
Jul 19 00:26:26 ER6-U35 kernel: zio pool=pool-655033360322456518 vdev=/dev/disk/by-id/wwn-0x5000cca26491428c-part1 error=5type=2 offset=498393575424 size=561152 flags=40080c80
```

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+vdev_disk_check_pages(abd_t *abd)
+{
+	vdev_disk_check_pages_t s = {0};
+	if (abd_iterate_func(abd, 0, abd_get_size(abd),
```
In FreeBSD similar code uses zio->io_size here. I am not completely sure it matches abd_get_size(abd) or is allowed to be smaller.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		struct page *page = is_vmalloc_addr(buf) ?
+		    vmalloc_to_page(buf) : virt_to_page(buf);
+
+		size_t take = MIN(len, page_size(page)-off);
```
Same page_size(page)-off as above.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -930,37 +1138,44 @@ vdev_disk_io_start(zio_t *zio)
+		ASSERT(zio->io_type == ZIO_TYPE_READ ||
+		    zio->io_type == ZIO_TYPE_WRITE ||
+		    zio->io_type == ZIO_TYPE_TRIM ||
+		    zio->io_type == ZIO_TYPE_IOCTL);
```
This looks like impossible condition. Looking on the previous cases.

robn replied:
Its supposed be, to make it trip in dev, but of course `ASSERT(0)` would have just done as well. Changed to `ASSERT3S(zio->io_type, ==, -1)` to get the type into the crash output.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	 * it can batch them if it wants.
+	 */
+	if (vbio->vbio_nbios > 1)
+		blk_start_plug(&plug);
```
It's not exported to modules.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	 * If we copied the ABD before issuing it, clean up and return the copy
+	 * to the ADB, with changes if appropriate.
+	 */
+	void *buf = abd_to_buf(vbio->vbio_abd);
+	abd_free(vbio->vbio_abd);
+	vbio->vbio_abd = NULL;
+
+	if (zio->io_type == ZIO_TYPE_READ)
+		abd_return_buf_copy(zio->io_abd, buf, zio->io_size);
+	else
+		abd_return_buf(zio->io_abd, buf, zio->io_size);
```
The abd is made from `abd_get_from_buf()` over the borrowed buf. That doesn't own the buf, so won't free it after the abd is freed.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+{
+	VERIFY0(atomic_read(&vbio->vbio_ref));
+
+	for (int i = 0; i < vbio->vbio_max_bios; i++)
+		if (vbio->vbio_bio[i])
```
Oh yeah good call.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		struct page *page = is_vmalloc_addr(buf) ?
+		    vmalloc_to_page(buf) : virt_to_page(buf);
+
+		size_t take = MIN(len, page_size(page)-off);
```
This took me down a deep rabbit hole, and the rework is much better for it.

A `struct page` for a compound page "tail" is basically just a regular page: 4K. So this works, but means if `buf` is not in the head page, then we're just loading regular pages in. And that's the case much of the time for larger allocations from the vmalloc slabs, so we lose.

So now for compound pages, we take the offset from the head instead, which has the size of the entire page. And so, compound pages are now loaded once and that's it. Lovely.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+vbio_alloc(zio_t *zio, struct block_device *bdev, unsigned int pages)
+{
+	unsigned int max_segs = vdev_bio_max_segs(bdev);
+	int max_bios = (pages / max_segs)+1;
```
The math was slightly off here, its true, but you got me thinking more about logical block sizes and yes, I hadn't managed it properly.

In another part of the code, just pushed, I now get the logical block size from the device, and use that when checking the alignment - instead of making sure they're all page aligned, I instead make sure they're all LBS aligned. Really only matters for the head and tail parts anyway, I think.

Linux up to 6.7 restricts logical block sizes to one of 512, 1024, 2048 or 4096 bytes.

robn commented:
[Last push](https://github.com/openzfs/zfs/compare/fbd8f5563e96180ecdc59c771ffe96a279d6a4b8..12c58b463a030e3d072fc58d58e7bc7c5d2d18b5) reworks a few things in response to review comments, and other things discovered while trying to answer them:

- always load compound pages from the head in one go
- check alignment against device logical block size, not page size
- only allocate as many segments in a BIO as we intend to fill
- use our own BIO count to free them, so we don't have to check if they exist
- use ZIO size instead of ABD size, just to be sure we're not doing too much
- make int types a bit more uniform
- minor cleanups

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,488 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	zio_delay_interrupt(zio);
+
+	/* Finish cleanup */
+	vbio_free(vbio);
```
Yeah, my thought was to get the zio into the pipeline at the earliest possible opportunity, but now I say it loud its a bit silly. I'll fix that up.

robn replied:
Tested doing the vbio cleanup before submitting the zio, and there's a small but measurable throughput drop on a couple of test workloads - about 1-2%. They're still ~2.6x over master so I won't be sad about losing them, but still.

I've rewritten a comment to try to make it clearer why its this way, but if you'd rather the simpler version, I'd be ok with that. Its https://github.com/robn/zfs/commit/270b44c49923c0a3c41e9bba390b9b6604db929c.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,488 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	 */
+
+	/* LBS is guaranteed to be one of 512, 1024, 2048 or 4096 */
+	uint_t lbs = bdev_logical_block_size(bdev);
```
Yeah, not really documented, but its there:

https://github.com/torvalds/linux/blob/3b47bc037bd44f142ac09848e8d3ecccc726be99/block/bdev.c#L138-L146
https://github.com/torvalds/linux/blob/3b47bc037bd44f142ac09848e8d3ecccc726be99/drivers/scsi/sd.c#L2690-L2695

behlendorf replied:
Well not exactly.  There's no guarantee that the page size is 4K.  These days even RedHat ships a kernel with 64K pages for aarch64.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/managing_monitoring_and_updating_the_kernel/what-is-kernel-64k_managing-monitoring-and-updating-the-kernel

robn replied:
Interesting. That's about memory page size though, not disk block sizes? (not that its not important, I'm just making sure I understand your point).

Its a good point though; I should try to test this a bit on something that isn't x86_64, and see if I can weird up the memory setup.

behlendorf replied:
Right.  The concern is that on system configured with large pages it's possible to have a LBS > 4K.  I'm fairly sure that works today in master, but we should make sure and update the "LBS is guaranteed to be" comment above accordingly.

robn replied:
So as best I can tell, LBS >4K isn't possible yet: there's some work going on, but it hasn't landed yet. See https://kernelnewbies.org/KernelProjects/large-block-size ([archive.org link](http://web.archive.org/web/20230705052950/https://kernelnewbies.org/KernelProjects/large-block-size); its been down for a few days).

Large pages certainly are though! I suspect that this code will work as-is (or maybe with very slight adjustment), so long as we keep using `struct page` and `page_size()` and never `PAGE_SIZE` or similar. I have dug out an arm64 board, which I'll try to get going on 64K pages and see what happens.

robn replied:
Made some small comment adjustments and another battery of tests. It seems sensible.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,488 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	 * Note if we're taking less than a full block, so we can check it
+	 * above on the next call.
+	 */
+	s->end = len & s->bmask;
```
I could just leave it at PAGE_SIZE, but I found with a 512b LBS and pushing everything into tiny allocations (`metaslab_force_ganging=32768`, `metaslab_force_ganging_pct=100`) it was constantly finding gaps it couldn't resolve, and had to linearise. Granted, that's an extreme case, but it seemed no big deal to just use the block size.

Ironically, the kernel should basically never need to split a BIO anymore, because we're never overfilling it. I suppose it gets trickier once various DM layers are in place, as we saw with LUKS, but I doubt its gonna be that much of an issue.

> What Linux will do...

Something bonkers, I'm sure.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,488 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	    .end = 0
+	};
+
+	if (abd_iterate_func(abd, 0, size, vdev_disk_check_pages_cb, &s))
```
This is a good idea. I'll have a play with it.

robn replied:
I did play with it, and there's something there, but I haven't got it work properly yet. I'm outta time for play, and I wouldn't want to merge it with this PR anyway. I'll come back to it.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -930,37 +1213,41 @@ vdev_disk_io_start(zio_t *zio)
 		return;
 	}
+
+	__builtin_unreachable();
```
Yeah, there's no great way to write it. Compiler will complain if I leave out `default:`, and if I have nothing after the switch. At least this is explicit.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,475 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	/*
+	 * This page is part of a "compound page", which is a group of pages
+	 * that can be referenced from a single struct page *. Its organised as
+	 * a "head" page, followed by a serious of "tail" pages.
```
Its been a long week :dizzy_face:

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,126 +591,475 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+	uint_t npages = vdev_disk_check_pages(abd, zio->io_size, lbs);
+	if (npages == 0) {
+		void *buf;
+		/* We can't safely use pages from this ABD, so use a copy */
```
Nice, I wanted the same thing but didn't really know where to put it. Will do!

robn replied:
Added `borrow_copy_total` and `borrow_copy_size_total`. The `*_cnt` stats are counting what's alive right now, but borrows are too transient to usefully count them that way. Also only counting copies; just borrowing isn't really that big a deal for linear abds.

This isn't the most useful stat perhaps, because there's other places that borrows happen, but it should be ok for order-of-magnitude comparisons. Probably useful to add stats to the vdev itself sometime, but I won't do that right now.

robn commented:
I believe this is good to go, pending outcome from test results.

@behlendorf If/when this lands, I'd like to also take it to 2.2, but I'm wary of introducing a change so fundamental on a stable series. I'm thinking about a module load parameter that toggles which implement is used, default to the old one. And then we could recommend folk struggling with LUKS (#15533), or seeing rejected SCSI IO (original case), or just wanting to try it out, get more performance, to give it a try, without risk. Assuming the implementation is sane, would you be interested in that?

(Incidentally, I am dogfooding this on my daily driver now (on top of 2.2.2). Its a pretty boring laptop workload though, but its not nothing).

robn commented:
Alright, see what you think of that.

Changes to previous:
- no longer removing `abd_bio_map` etc
- re-lifed the old `dio` and `physio` stuff, pulled it into the same part of the file, renamed `vdev_classic_*` but no implementation changes
    - not even in the "computer BIO segments" function, that is, `zfs_vdev_disk_max_segs` is ignored for classic
- added `zfs_vdev_disk_classic` tuneable; `0` for "new", `1` for "classic", default 0
    - changeable at module load time only (runtime is quite easy, but seems bonkers)
    - if you set it explicitly, it'll put something on the kernel log as a debugging assist
    - doc advises user to send a bug report if they use it

I've done as much testing as I can on both modes. I managed to induce a split (and SCSI rejection with "classic"), none on "new", as expected. Performance on my sanity check workload remains 2.7x on the new.

sempervictus commented:
@robn - how safe is this to test on systems where i'd need to restore data if i blow things up? :smile:
Far as the page alignment comment in vdev_disk.c - looks like we stumbled on something in the UIO code which tries to map more than one page at a time by miscalculating the size passed to `kmap`; so think that guarding/testing for mm problems on the target OS/arch is a great idea.

robn commented:
Top commit adds `abd_iterate_page_func`, which yield `struct page` instead of data buffers, handling compound pages as it goes. Since the BIO gets filled with `struct page`s directly, this obviates the need to map every page along the way.

@behlendorf I'll be interested to know if that changes anything in your test case.

robn commented:
@sempervictus I think it just works or doesn't - I can't think of any reason why a bug would cause any on-disk corruption. So worst case, you should be able to switch it over to `zfs_vdev_disk_classic=1` or just roll back to an earlier version.

amotin reviewed (commented):
I hope we drop zfs_vdev_disk_classic and the duplication pretty soon.  I would not call old code "very stable", considering we know it has issues with #15452.

robn commented on include/sys/abd_impl.h:
```diff
@@ -51,6 +53,10 @@ struct abd_iter {
+
+	struct page	*iter_page;		/* current page */
+	size_t		iter_page_dsize;	/* size of data in page */
+	size_t		iter_page_doff;		/* offset of data in page */
```
The union-of-structs is a good idea, pushed.

Incidentally, I've started some cleanup to properly make the OS-specific parts be out in `include/os/xxx/sys/abd_os.h`, just like we do everywhere else. I won't hold up this PR for it as I like won't finish it until January, but thought I should mention that its coming.

robn commented:
> I hope we drop zfs_vdev_disk_classic and the duplication pretty soon. I would not call old code "very stable", considering we know it has issues with #15452.

Maybe "very stable" is too far, but at least its a known quantity that has served well for a long time. I would hope that the new code can get enough testing on 2.2 that we could remove classic for 2.3.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,8 +590,539 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		    "Illegal access %llu size %llu, device size %llu",
+		    (u_longlong_t)io_offset,
+		    (u_longlong_t)io_size,
+		    (u_longlong_t)i_size_read(bdev->bd_inode));
```
I know it is not your code, but why do we compare value fetched directly, while printing more elaborately?

robn replied:
Actually I don't know. Worse, I don't even know how this compiled - `io_offset` and `io_size` are not even declared here post-refactor! I've fixed that up (made them `zio->io_*`) but I'll leave it for now and try to clean up more later.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -670,8 +590,539 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+vbio_return_abd(vbio_t *vbio)
+{
+	zio_t *zio = vbio->vbio_zio;
+	if (vbio->vbio_abd == NULL || vbio->vbio_abd == zio->io_abd)
```
Isn't the second condition excessive after being checked in vdev_disk_io_rw() on line 988?

robn replied:
Yeah, good spot. It was left over from a previous iteration. Removed.

robn commented on include/sys/abd_impl.h:
```diff
@@ -38,12 +38,30 @@ typedef enum abd_stats_op {
+			/* size of data in page */
+			size_t		iter_page_dsize;
+			/* offset of data in page */
+			size_t		iter_page_doff;
```
Done.

robn commented on module/os/linux/zfs/abd_os.c:
```diff
@@ -897,6 +928,9 @@ abd_iter_init(struct abd_iter *aiter, abd_t *abd)
 	}
+	aiter->iter_page = NULL;
+	aiter->iter_page_doff = 0;
+	aiter->iter_page_dsize = 0;
```
Yeah, good spot. I've done `bzero` (well, `memset`) instead.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -63,16 +71,7 @@ static uint_t zfs_vdev_open_timeout_ms = 1000;
-	int			dr_bio_count;	/* Count of bio's */
-	struct bio		*dr_bio[];	/* Attached bio's */
-} dio_request_t;
+
```
Heh, fair.

robn commented on module/zfs/abd.c:
```diff
@@ -696,6 +697,7 @@ abd_return_buf_copy(abd_t *abd, void *buf, size_t n)
 {
 	if (!abd_is_linear(abd)) {
 		abd_copy_from_buf(abd, buf, n);
+		abd_update_borrow_stats(abd);
```
I'm sort of indifferent at this point. I thought it would be more useful than it was, however it was useful to spot a bug on ARM, but only because it was a couple of orders of magnitude different to what it should have been.

I do want to do a more comprehensive stats package for `vdev_disk` proper, and this state probably is better done there.

@behlendorf you suggested this earlier. Would you rather keep it or remove it? I am fairly ambivalent.

behlendorf replied:
I wouldn't object to simply dropping this.  I also initially thought it would be more useful.

robn replied:
And there it goes... :balloon:

robn commented on include/sys/abd_impl.h:
```diff
@@ -38,12 +38,30 @@ typedef enum abd_stats_op {
+		/* for abd_iter_page() */
+		struct {
+			/* current page */
+			struct page	*iter_page;
```
Ahh damn - the curse of late changes strikes again!

Yeah, I think your change is ok. Its at least defensible in that it follows setup->call->cleanup pattern in `abd_iterate_func()`. You're right about the asserts being unclear, and we could just add asserts for `iter_page*` as well, just to make it clear "all the iterator state should be zero", even though its a bit stupid.

robn replied:
Last push adds your patch, and some extra asserts for them, and comment.

behlendorf commented:
Almost there.  This has been working well in my testing, but it appears the CI uncovered an issue on [CentOS 7](https://build.openzfs.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/1561/steps/shell_4/logs/console).  I've resubmitted this build to get a second test run.

robn commented:
Yuck. I have a CentOS 7 VM here; I'll see what I can discover. Might not be until tomorrow morning though (~20h from now).

robn commented:
Very first IO:

```
[  995.632381] VERIFY3(vbio->vbio_cur < vbio->vbio_max_bios) failed (1 < 1)
[  995.632909] PANIC at vdev_disk.c:707:vbio_add_page()
[  995.633429] Showing stack for process 24505
[  995.633433] CPU: 0 PID: 24505 Comm: z_null_iss Tainted: P           OE  ------------   3.10.0-1160.90.1.el7.x86_64 #1
[  995.633434] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
[  995.633435] Call Trace:
[  995.633442]  [<ffffffff929b1bec>] dump_stack+0x19/0x1f
[  995.633449]  [<ffffffffc061e65b>] spl_dumpstack+0x2b/0x40 [spl]
[  995.633454]  [<ffffffffc061e739>] spl_panic+0xc9/0x110 [spl]
[  995.633458]  [<ffffffff923cc275>] ? mempool_alloc_slab+0x15/0x20
[  995.633460]  [<ffffffff923cc3ee>] ? mempool_alloc+0x6e/0x170
[  995.633549]  [<ffffffffc08a4912>] ? abd_iter_advance+0x72/0x1b0 [zfs]
[  995.633553]  [<ffffffff9249b8a9>] ? bvec_alloc+0x59/0x120
[  995.633555]  [<ffffffff9249bb83>] ? bio_alloc_bioset+0x213/0x310
[  995.633629]  [<ffffffffc08a7b86>] vbio_add_page+0xe6/0x180 [zfs]
[  995.633693]  [<ffffffffc08a7c20>] ? vbio_add_page+0x180/0x180 [zfs]
[  995.633756]  [<ffffffffc08a7c39>] vdev_disk_fill_vbio_cb+0x19/0x20 [zfs]
[  995.633805]  [<ffffffffc06f1fa8>] abd_iterate_page_func+0xd8/0x1a0 [zfs]
[  995.633869]  [<ffffffffc08a7e2d>] vdev_disk_io_rw+0x12d/0x240 [zfs]
[  995.633931]  [<ffffffffc08a800b>] vdev_disk_io_start+0xcb/0x4a0 [zfs]
[  995.633994]  [<ffffffffc0896f3d>] zio_vdev_io_start+0x1ed/0x560 [zfs]
[  995.634057]  [<ffffffffc088c180>] zio_execute+0xf0/0x220 [zfs]
[  995.634063]  [<ffffffffc0626979>] taskq_thread+0x2f9/0x4b0 [spl]
[  995.634067]  [<ffffffff922e1170>] ? wake_up_state+0x20/0x20
[  995.634130]  [<ffffffffc088c090>] ? zio_taskq_member.constprop.0+0x80/0x80 [zfs]
[  995.634135]  [<ffffffffc0626680>] ? taskq_thread_spawn+0x60/0x60 [spl]
[  995.634139]  [<ffffffff922cb621>] kthread+0xd1/0xe0
[  995.634141]  [<ffffffff922cb550>] ? insert_kthread_work+0x40/0x40
[  995.634144]  [<ffffffff929c51f7>] ret_from_fork_nospec_begin+0x21/0x21
[  995.634146]  [<ffffffff922cb550>] ? insert_kthread_work+0x40/0x40
```

I'll have to put a bit more debug on it. I expect it'll be `vdev_bio_max_segs()` gone wrong. Tonight or tomorrow probably.

robn commented:
Right, I think I understand it. Its annoying but I think shows up a deficiency.

BIOs have two limits: number of segments, and max size. On the real devices I've seen the max size is up around UINT32_MAX, but also a light code read suggests to me that the the block layer in later kernels will automatically split BIOs that are too big by size.

In the EL7 kernel, loop devices appear to have max size 64K, and the splitting isn't automatic, but up to the block driver. So we're blowing the BIO size, even though it has free segments in it.

I could do a couple of dumb things to workaround it but the whole point of this PR is to not require dumb workarounds anymore. So I think I would like to pause this for a moment, and I'll rework the up-front page checking stage to count segment sizes, considering BIO limits, and get that math right.

I don't think its a big or onerous job, but I am about to check my brain out for a couple of weeks so its likely to be second week of January at the latest before I can get back to this (unless I decide to fiddle with it on the couch while watching cricket, which is likely, but I don't want to commit to it).

(I'll be a tiny bit sad when we drop support for EL7, as its found some interesting bugs. I'll mostly be happy though 🫠).

robn commented:
Main change in last push is to track number of bytes used when scanning the ABD up front. If adding the current page would blow the BIO byte limit, we just bump the segment count by the number of segments remaining in the BIO, effectively pushing that page into the next BIO. This matches what the fill callback will do, and so we end up creating enough BIOs.

This is enough to fix the trivial crash on CentOS 7. I've run my sanity and aggressive-ganging and perf checks and everything seems to be working fine. We'll see what the tests shake out.

robn commented:
I also reordered the commits to put the ABD page stuff first, so I could collapse the original version with the updated one, since it was just a mess.

I'm thinking about reorganising a little more so that I don't remove and then re-add the "classic" version, as its making a mess of the git history. In that case I'd likely do the function rename and maybe the selector in an initial commit, to keep the "rewrite" commit as tight as possible. But maybe it doesn't matter if we're planning to remove this code outright in a few months (?) anyway.

robn commented:
Bummer. Still, lots further. I'll check it out soon!

robn commented:
Bummer. Still, lots further. I'll check it out soon!

robn commented:
> Bummer. Still, lots further. I'll check it out soon!

Got it, just a counting bug.

robn commented:
Last push reorganises the commits to keep the individual diffs cleaner. The intermediate stages are mostly unuseful on their own (though they compile and work), but they show a logical progression. The final diff is identical to previous, apart from one function moving, which was a mistake in the original anyway.

@behlendorf assuming tests pass, I believe this is ready to go.

behlendorf commented:
@robn the reorganized patch stack works for me.  Unfortunately, it looks like we're still hitting some rough edges on [CentOS 7](https://build.openzfs.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/1690/steps/shell_4/logs/console).  Aside from that this looks good.

robn commented:
Alright, that _should_ be everything.

The EL7 crash logs showed two issues. The VERIFY trip was because of a miscount, because it turns out you can't trust `bi_vcnt` as "number of segments I have added to this BIO", so the math was coming out wrong. I'm not sure why you can't, I just replaced it with my own counter. Fine.

The second thing though led to a significant rework, and then an un-rework, which I will explain.

Through the logs, we see:

```
Jan  9 17:59:28 centos7 kernel: bio too big device loop0 (256 > 255)
Jan  9 17:59:28 centos7 kernel: zio pool=testpool vdev=/dev/loop0 error=5 type=1 offset=24717312 size=131072 flags=1605809
```

`checksum/filetest_001_pos` was hitting this, among others.

On EL7, loop devices have `max_sectors` of 255. This is the byte limit for a single BIO, that is, 127.5KiB.

After adding some debug, it turns out that the ABD inside this zio is a single 128KiB page, that is, an `order=5` compound page. The new code doesn't break pages, so all it could do was load the entire page into one BIO.

The EL7 kernel requires the caller to correctly fill the BIO, honouring all limits. There's functions available to request certain transformations (like splitting) but its all on the caller. More recent kernels on the other hand mostly don't care what you do, and will check the BIO when its submitted and make whatever transformations it wants to get it down under the limits.

(I guess this change happened somewhere in the 4.x series, but I haven't tried to find out.

The code just pushed goes back to the less complex pre-new-year code, with one addition: when filling the BIO, we check if the data in the page we're about to load will fit, that is, be under the byte limit. If not, we load as much as will fit, maintaing LBS alignment, and then put the rest into the next BIO. That is, we do our own limited form of page splitting, an irony that is not lost on me.

For recent kernels, that means we're doing splitting work that the kernel was already doing, possibly better than us. This shouldn't be a big deal; we're preserving alignment, which is all we care about. If there's something else wrong, it will adjust over it. For EL7, we're ensuring we stay under segment and byte limits in each BIO. There don't seem to be any other limits of note, so I'm hopeful this is enough.

The rework I mentioned goes like this.

When you call `bio_alloc()`, you have to pass in the count of segments you intend to add. This controls the size of the `struct bio_vec []` array inside the bio; each `bio_vec` is 16 bytes. There's also an optimisation; the `struct bio` itself has 4 `bio_vec`s directly inside it; if you request <= 4, you don't take an additional allocation.

Previously, the up-front alignment checking code counted pages as part of its work, and this count was used as the segment count to `bio_alloc()`. Assuming 1:1 match of ABD pages to segments loading into the BIO, this is great. As soon as we have to start considering byte limits though, then we lose this 1:1 match (we might underfill a BIO, or have to split a single page across two segments in two BIOs).

This gives us two options, neither great. We can either include the idea of underfilling or splitting into the up-front checking code (effectively a "simulated fill"), or we can always request the maximum possible number of segments in every BIO, even if we only need a small handful of them. I tried the first option [here](https://github.com/robn/zfs/blob/32e0411d23f31a644d039395bc9411b54919f407/module/os/linux/zfs/vdev_disk.c) but it ends up meaning that the check and the fill functions repeat each other as well as being quite complex. So I've opted for the second version in this PR, because its much easier to follow.

There is a nice bit though. Because we don't know how up front how many BIOs we need, I switched the store of completed BIOs to be a linked list instead of an array. It looks like most users in the kernel do this, so it seems reasonable. Because of that, we don't need to allocate the array of BIO pointers, so its just a single `vbio_t` allocation. The seglist array inside the BIO comes from a memory cache, so most allocations are probably going to reuse cache allocations anyway, so I suspect it ends up just moving allocations around rather that adding or removing overheads.

So! As I write this, its showing the same performance profile as all the previous versions, which is nice, and also the tests that have been failing/crashing on EL7 are passing. However there appears to be some kind of memory corruption that I'm struggling to track down. its most easily reproduced by with the `pool_checkpoint/checkpoint_removal` test; on my local toy VMs it results in the loop driver throwing IO errors, but on a real computer I got multiple CPUs softlocked, which is almost always some memory corruption. I'll dig into that more soon, but I expect it'll be something very stupid.

robn commented:
Pretty sure I figured it all out. There was a real infinite loop, I screwed up handling zero space left in the BIO. The rest was `checkpoint_removal` apparently requiring a lot more than 2G of RAM in the VM to work.

Perf tests good, my local test suite run is progressing, a handful of tests troublesome tests on EL7 also looking good, and I've just put the whole lot onto my workstation and laptop as daily driver. If CI likes it and code read doesn't turn up anything weird, then this might just be there!

robn commented:
EL7 failures are a total system hang during `zpool_reopen/zpool_reopen_003_pos`. I did get a kernel crash dump on a local run:

```
Jan 12 18:34:36 centos7 kernel: ------------[ cut here ]------------
Jan 12 18:34:36 centos7 kernel: WARNING: CPU: 0 PID: 2311 at lib/kobject.c:658 kobject_put+0x50/0x60
Jan 12 18:34:36 centos7 kernel: kobject: '(null)' (ffff986bbb68f038): is not initialized, yet kobject_put() is being called.
Jan 12 18:34:36 centos7 kernel: Modules linked in: scsi_debug loop zfs(POE) spl(OE) ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 ipt_REJECT nf_reject_ipv4 xt_conntrack ebtable_nat ebtable_broute bridge stp llc ip6table_nat nf_conntrack_ipv6 nf_defrag_ipv6 nf_nat_ipv6 ip6table_mangle ip6table_security ip6table_raw iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat iptable_mangle iptable_security iptable_raw nf_conntrack ip_set ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter sunrpc ext4 mbcache jbd2 intel_pmc_core intel_powerclamp iosf_mbi crc32_pclmul ghash_clmulni_intel ppdev aesni_intel sg lrw gf128mul pcspkr glue_helper ablk_helper cryptd parport_pc parport video i2c_piix4 binfmt_misc ip_tables xfs libcrc32c sd_mod crc_t10dif crct10dif_generic ata_generic pata_acpi vmwgfx drm_kms_helper
Jan 12 18:34:36 centos7 kernel: syscopyarea sysfillrect sysimgblt fb_sys_fops ttm ata_piix ahci libahci drm crct10dif_pclmul crct10dif_common crc32c_intel serio_raw e1000 drm_panel_orientation_quirks libata dm_mirror dm_region_hash dm_log dm_mod
Jan 12 18:34:36 centos7 kernel: CPU: 0 PID: 2311 Comm: vdev_open Tainted: P           OE  ------------   3.10.0-1160.90.1.el7.x86_64 #1
Jan 12 18:34:36 centos7 kernel: Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
Jan 12 18:34:36 centos7 kernel: Call Trace:
Jan 12 18:34:36 centos7 kernel: [<ffffffffad5b1bec>] dump_stack+0x19/0x1f
Jan 12 18:34:36 centos7 kernel: [<fffffffface9f868>] __warn+0xd8/0x100
Jan 12 18:34:36 centos7 kernel: [<fffffffface9f8ef>] warn_slowpath_fmt+0x5f/0x80
Jan 12 18:34:36 centos7 kernel: [<ffffffffad09e2d6>] ? bdev_inode_switch_bdi+0x66/0xf0
Jan 12 18:34:36 centos7 kernel: [<ffffffffacec18f1>] ? try_to_grab_pending+0xc1/0x170
Jan 12 18:34:36 centos7 kernel: [<ffffffffad19f100>] kobject_put+0x50/0x60
Jan 12 18:34:36 centos7 kernel: [<ffffffffad2d3ca7>] put_device+0x17/0x20
Jan 12 18:34:36 centos7 kernel: [<ffffffffad09f0ac>] __blkdev_put+0xac/0x1b0
Jan 12 18:34:36 centos7 kernel: [<ffffffffad09fb2c>] blkdev_put+0x4c/0x140
Jan 12 18:34:36 centos7 kernel: [<ffffffffc0bee973>] vdev_blkdev_put+0x23/0x30 [zfs]
Jan 12 18:34:36 centos7 kernel: [<ffffffffc0beffe0>] vdev_disk_open+0x220/0x580 [zfs]
Jan 12 18:34:36 centos7 kernel: [<fffffffface2c5ee>] ? __switch_to+0xce/0x590
Jan 12 18:34:36 centos7 kernel: [<ffffffffc0b53dc0>] vdev_open+0x1a0/0x9b0 [zfs]
Jan 12 18:34:36 centos7 kernel: [<ffffffffc0b54762>] vdev_open_child+0x22/0x40 [zfs]
Jan 12 18:34:36 centos7 kernel: [<ffffffffc096ca59>] taskq_thread+0x2f9/0x4b0 [spl]
Jan 12 18:34:36 centos7 kernel: [<ffffffffacee8619>] ? pick_next_entity+0xa9/0x190
Jan 12 18:34:36 centos7 kernel: [<ffffffffacee1170>] ? wake_up_state+0x20/0x20
Jan 12 18:34:36 centos7 kernel: [<ffffffffc096c760>] ? taskq_thread_spawn+0x60/0x60 [spl]
Jan 12 18:34:36 centos7 kernel: [<ffffffffacecb621>] kthread+0xd1/0xe0
Jan 12 18:34:36 centos7 kernel: [<ffffffffacecb550>] ? insert_kthread_work+0x40/0x40
Jan 12 18:34:36 centos7 kernel: [<ffffffffad5c51f7>] ret_from_fork_nospec_begin+0x21/0x21
Jan 12 18:34:36 centos7 kernel: [<ffffffffacecb550>] ? insert_kthread_work+0x40/0x40
Jan 12 18:34:36 centos7 kernel: ---[ end trace 8b3dc4ca130d6717 ]---
```

I should be able to look at it this evening.

(newer kernels must be very forgiving of accounting mistakes, or older ones are very unforgiving. Either way, this suuuucks).

Rudd-O commented:
This is awesome progress!  Thanks.

robn commented:
Short answer is that I'm corrupting memory somewhere in the BIO cleanup stage. I don't know if its a 3.x thing or a EL7 thing, but something is different in that very old kernel and I'm struggling to pin it down.

I've pushed an early change to sidestep this entirely by chaining the group of BIOs together such that the kernel only has to post a single completion for the entire set. It will take care of the cleanup of everything but the head, and we don't need to track or refcount them at all.

Initial tests look promising. I've pushed it now to let the runners have a look at it. If it looks good, I'll clean it up we'll use it. It could be a good change either way; simpler code.

robn commented:
No change, sigh. At least that rules out misuse of `bio_put()`. `zpool_reopen/zpool_reopen_003_pos` is a reliable trigger, so my next act is to go through it step by step and try to understand the timing.

(this was never suppose to be this complicated!)

robn commented:
I've been working on this sporadically for the last month. Definitely made progress, though not solved yet.

The `bio_chain()` mechanism is very good, and simplifies our code - no need to refcount and track multiple BIOs, just the one at the head of the list. So I've tidied up that code.

On EL7 though, it still brings down the machine. I've enough apparently-unrelated crash dumps to feel confident that there's definitely a memory corruption happening, and since its still not coming up outside of the reopen test, it has to be something to do with device vs IO lifetime.

I've not been able to reproduce on any stock kernel yet, but have not been able to test back to earlier than 3.18.

I'll keep pushing on it. As I said earlier, I'd be happy to just drop EL7 support if it turns out its only being weird there, but I need to understand it first, just in case I'm doing something wrong that later kernels are able to deal with. This approach has already paid off - I was calling `bio_chain()` with the args reversed, which newer kernels apparently didn't mind. So that was a good one to fix.

More in a while I guess!

robn commented:
Alright, I'm kinda baffled at this point. I've narrowed things down a long way but I can't come up with a workable theory.

The test in question is `cli_root/zpool_reopen/zpool_reopen_003_pos.ksh`. It creates a two-disk mirror, (virtually) pulls a disk to degrade it, writes some data, and starts a scrub. Then it returns the disk and checks that the scrub is replaced by a resilver.

To ensure the scrub is running when the disk is returned, it uses `zinject` to add a 25ms delay to the remaining pool. Once the disk is returned and the resilver checked, `zinject -c all` is used to clear the injection records.

Note the ordering:
- pool created `loop1` `sdb1`
- `sdb1` removed
- data written
- `loop1` delay injected
- scrub started
- `sdb1` restored
- pool reopened
- wait for resilver
- remove injected delay

I found that if the delay injection is never removed (comment out `zinject -c all`), the test runs to successful completion. So its not the presence of the delay injection alone that causes the problem.

If I change `vbio_completion()` to call `zio_interrupt()` directly instead of `zio_delay_interrupt()`, the test completes. This makes sense, as there is where the delay is applied.

If I modify `zio_delay_interrupt()` to never dispatch the zio to the taskq, the test completes. That makes sense too; its effectively the same as the the delay never being applied.

If I remove the taskq dispatch and instead always delay inline (ie always go through `zfs_sleep_until()`), we get the same behaviour as with the taskq: the stock test corrupts memory, but removing `zinject -c all` makes the test run to completion. So the delay mechanism is not implicated in any way.

But as noted right at the start, the delay itself is not the problem, but rather, removing the delay during the test. Given that, the only thing that makes any sense to me is that there's its related to ordering:

- delay injected
- IO1 issued, completes. response processing is delayed
- delay injection removed
- IO2 issued, completes, response processed
- IO1 response processed

And there's some dependency in there, such that IO2's response processing frees resources that IO1's response processing needs.

_Except_, that this doesn't happen using the "classic" `vdev_disk` code, and it only happens on EL7. This suggests something subtly different in memory management - are scatter pages or larger slab allocations reclaimed "immediately" or more frequently on EL7?

The new code doesn't seem that subtle or sensitive though. I've read it all over yet again, and especially trying to map out resource lifetime, and I still can't see anything that looks off. I must be doing wrong somewhere, but I just can't see it.

I'm glad to have narrowed it down somewhat. I'll keep trying to reduce it further and remove more variables. Definitely a weird one!

Question for @behlendorf: 745ace3f2 (#10489) puts `zpool_reopen_003_pos` on the "fails sometimes" list, but doesn't say why. Do you remember anything about it? I'm wondering if I'm actually running into a "known" issue.

chrisrd commented:
Hi Rob,

Can you clarify this part?

> If I remove the taskq dispatch and instead always delay inline (ie always go through
> zfs_sleep_until()), we get the same behaviour as with the taskq: the stock test corrupts memory, but
> removing zinject -c all makes the test run to completion. So the delay mechanism is not implicated
> in any way.

Do you mean that without the `zinject -c all` the test succeeds, but there's still corrupted memory (detect how?)? Or, without the `zinject -c all` the test succeeds and there's no apparent corrupted memory?

It sounds like you're tripping over some pre-existing problem with the zinject infrastructure. If we're getting memory corruption due to the `zinject -c all` it sounds like _something_ is corrupting the list used to keep track of the injections.

Quite possibly barking up the wrong tree, but one thing that struck me as odd in that area is that there are two different definitions of `list_node_t` used within `inject_handler_t` used by the zinject list:

`module/zfs/zio_inject.c`
````
typedef struct inject_handler {
        int                     zi_id;
        spa_t                   *zi_spa;
        zinject_record_t        zi_record;
        uint64_t                *zi_lanes;
        int                     zi_next_lane;
        list_node_t             zi_link;              <<<<
} inject_handler_t;
````
The first definition of `list_node_t` (in current master) is:

`include/os/linux/spl/sys/list.h`
````
typedef struct list_head list_node_t;

typedef struct list {
        size_t list_offset;
        list_node_t list_head;
} list_t;
````

Note: commit 78e8c1f844cb5 (2024-01-26) changed this from:
````
typedef struct list {
        size_t list_size;
        size_t list_offset;
        list_node_t list_head;
} list_t;
````

Then `list_head` comes from:

`linux/include/linux/types.h`
````
struct list_head {
        struct list_head *next, *prev;
};
````

The second definition of `list_node_t` is:

`lib/libspl/include/sys/list.h`
````
typedef struct list_node list_node_t;
````

`lib/libspl/include/sys/list_impl.h`
````
struct list_node {
        struct list_node *next;
        struct list_node *prev;
};
````

I.e. in the first definition the `list_t` has one (or two, depending on which version you're using) extra structure member.

I wonder if something about the (old) EL7 kernel is causing the wrong (or mixed?) definitions to be used? That could certainly cause memory corruption.

Cheers,

Chris

behlendorf commented:
That's downright mysterious.  Offhand, I can't think of anything that different with EL7 that would explain this.

> puts zpool_reopen_003_pos on the "fails sometimes" list, but doesn't say why. Do you remember anything about it? I'm wondering if I'm actually running into a "known" issue.

Unfortunately, I don't remember any details about this one.  It may be a long standing issue that just never got properly investigated.

robn commented:
Little breakthrough today. If I comment out the "move back compound head" section, everything is solid. Turns out before ~4.5, head and tail pages are refcounted separately (see torvalds/linux@ddc58f2). So I think what needs to happen is that when we move back to the head, we need to take a reference to it, and then later, after the IO is returned, release that reference.

I'm still confirming this, and figuring out the best way to track the reference that need to be released, as there's a couple of other things in play that might be a factor, for example gang child ABDs can reuse the contents of another ABD. I expect to have it sorted tomorrow, and hopefully nothing else comes up after that!

What a bonkers time! :sweat_smile:

robn commented:
If this turns out to be true, then my rough guess at what's happening is that the head page is entered into the bio, which takes a reference. After the bio completes, the last reference is dropped, and the head page is flagged for free, but not freed yet. If no delay is injected, then the zio is executed pretty much immediately, before the head page can be freed, but with the delay there, it gets freed before the zio is executed, and the done callbacks, whatever they are, start messing around in who-knows-what memory. And then the whole reopen thing is actually a red herring; there's just enough going on inside the kernel to trigger the page free (I'm guessing some RCU schedule).

I'm sure there's plenty of reasons why this isn't exactly right, but it feels ok as a working shape. I'm good with it for now!

robn commented:
Last commit takes care of it, and I see its passing on EL7 :tada: . I'll look at those 22.04 failures in the morning; I haven't had anything fail locally on multiple kernels so hopefully its nothing too big or just some transient failure.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -697,8 +614,460 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		 * math in vdev_disk_check_pages() is wrong. In either case,
+		 * something in seriously wrong and its not safe to continue.
+		 */
+		VERIFY(vdev_disk_check_pages(abd, zio->io_size, bdev));
```
I would use ASSERT() here.  It should not happen randomly, while the extra page traversal is not exactly trivial.

robn replied:
I'm torn on this. I take your point, but doing this ensures that all ABDs are checked before submission regardless of where they came from, and since confusion about page layout is where this PR started from, I feel like checking everything is a good hedge against future problems. This is also already the "slow path"; we shouldn't be here most of the time, so I feel like the additional overhead is worth it for the safety.

I think I'm pulling weakly towards leaving it as-is, until and unless it gets shown to be a hot spot. If you felt strongly in the other direction, then I'll change it. There's also maybe a middle road, where I put the same shape of checks directly into `vbio_fill_cb()`, making it marginally heavier but avoiding a second traversal. Let me know if you think I've got my risk assessment all wrong :)

amotin replied:
I probably agree that more time in this case will likely be spent on memory copying rather this traversal.  But I just can't imagine how the freshly allocated virtually contiguous buffer may not fit the requirements.  I don't see it as a show-stopper now, if you worry, but at very least we could scrap it after some time together with "classic" code.

robn replied:
Thanks for humoring me :) I'm cool with downgrading it once its been in the wild a little bit and especially if it solves weird issues that people see on LUKS etc. Cheers!

amotin reviewed (approved):
Looks good to me, just few minor nits.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -697,8 +614,460 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		 * overhead on the new.
+		 */
+		ssize = MIN(size, (vbio->vbio_max_bytes - BIO_BI_SIZE(bio)) &
+		    ~(vbio->vbio_lbs_mask));
```
Nice, taken.

robn commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -697,8 +614,460 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+#endif
+}
+
+static inline uint_t
+vdev_bio_max_bytes(struct block_device *bdev)
```
I'll leave it as-is. It matches all the other helpers, and being explicit aids readability.

robn commented:
> I'll look at those 22.04 failures in the morning; I haven't had anything fail locally on multiple kernels so hopefully its nothing too big or just some transient failure.

Looks like 22.04 failures are unrelated; something in the systemd scripts maybe, and other PRs are doing it too. Pretty sure not my fault.

@behlendorf pretty sure we're good to go! _[exhales three months of tension]_

robn commented:
> @chrisrd I wonder if something about the (old) EL7 kernel is causing the wrong (or mixed?) definitions to be used? That could certainly cause memory corruption.

Sorry I didn't reply to this! I was going to follow up but went off in another direction and forgot.

So the specific thing you saw isn't it; libspl is the userspace "version" of ZFS (ie what's used in zdb, ztest, etc). It provides its own implementation of `list_t` separate from the one we use in the kernel (`module/os/linux/spl/...`). But it did cause me to wonder if some other structure shared between the userspace zinject tool and the fault injection stuff in the kernel module where being filled out differently, and I spent a bunch of time looking at it. I decided in the end that no, there's nothing there, all is well. Regardless, it was a great suggestion, and I learned a lot more about zinject from it.

So thanks for the idea, appreciated! :)

chrisrd commented:
@robn I'm sorry to have taken up so much of your time on a dead end. That head/tail refcounting change, associated with the timing dependent bug trigger, looks like a real doozy to track down. I'm super impressed with the tenacity and insight you've brought to this issue - many, many thanks!

behlendorf reviewed (approved):
@robn thanks for sticking with this one and getting to the root cause!
